### PR TITLE
Check for NaN in gen_candidates_scipy

### LINKED ISSUE #685: Handling NaN inputs in acquisition functions
## Issue description

Sometimes, when optimizing an acquisition function with `optimize_acqf`, the solution becomes a tensor of `NaN`s, at which the acquisition function gets evaluated (haven't diagnosed what leads to `NaN` yet). This does not raise any errors until the posterior `rsample(...)` call (using qEI). An example stack trace is as follows:

```
/Users/saitcakmak/.conda/envs/other-BO/lib/python3.8/site-packages/gpytorch/lazy/lazy_tensor.py:1354: NumericalWarning: Runtime Error when computing Cholesky decomposition: Matrix not positive definite after repeatedly adding jitter up to 1.0e-04. Original error on first attempt: cholesky_cpu: For batch 8: U(7,7) is zero, singular U.. Using RootDecomposition.
  warnings.warn(
/Users/saitcakmak/.conda/envs/other-BO/lib/python3.8/site-packages/gpytorch/lazy/lazy_tensor.py:1354: NumericalWarning: Runtime Error when computing Cholesky decomposition: cholesky_cpu: 1000 of 1000 elements of the torch.Size([10, 10, 10]) tensor are NaN.. Using RootDecomposition.
  warnings.warn(
Traceback (most recent call last):
  File "/Users/saitcakmak/.conda/envs/other-BO/lib/python3.8/site-packages/botorch/posteriors/gpytorch.py", line 87, in rsample
    samples = self.mvn.rsample(
  File "/Users/saitcakmak/.conda/envs/other-BO/lib/python3.8/site-packages/gpytorch/distributions/multitask_multivariate_normal.py", line 230, in rsample
    samples = super().rsample(sample_shape=sample_shape, base_samples=base_samples)
  File "/Users/saitcakmak/.conda/envs/other-BO/lib/python3.8/site-packages/gpytorch/distributions/multivariate_normal.py", line 173, in rsample
    covar_root = covar.root_decomposition().root
  File "/Users/saitcakmak/.conda/envs/other-BO/lib/python3.8/site-packages/gpytorch/utils/memoize.py", line 59, in g
    return _add_to_cache(self, cache_name, method(self, *args, **kwargs), *args, kwargs_pkl=kwargs_pkl)
  File "/Users/saitcakmak/.conda/envs/other-BO/lib/python3.8/site-packages/gpytorch/lazy/lazy_tensor.py", line 1364, in root_decomposition
    evals, evecs = self.symeig(eigenvectors=True)
  File "/Users/saitcakmak/.conda/envs/other-BO/lib/python3.8/site-packages/gpytorch/utils/memoize.py", line 59, in g
    return _add_to_cache(self, cache_name, method(self, *args, **kwargs), *args, kwargs_pkl=kwargs_pkl)
  File "/Users/saitcakmak/.conda/envs/other-BO/lib/python3.8/site-packages/gpytorch/lazy/lazy_tensor.py", line 1596, in symeig
    return self._symeig(eigenvectors=eigenvectors)
  File "/Users/saitcakmak/.conda/envs/other-BO/lib/python3.8/site-packages/gpytorch/lazy/lazy_tensor.py", line 1879, in _symeig
    evals, evecs = torch.symeig(self.evaluate().to(dtype=torch.double), eigenvectors=eigenvectors)
RuntimeError: symeig_cpu: For batch 0: the algorithm failed to converge; 9 off-diagonal elements of an intermediate tridiagonal form did not converge to zero.
```  

I personally read the stack trace bottom up, so I completely miss the warning about `NaN`s until much later. I trace it step by step, originally speculating that it is an issue with the model etc, until I realize that the `forward` call was done with a tensor of `NaN`s. 

This is partially my fault for not looking through the stack trace more carefully, but I think this whole process can be made more efficient by catching `NaN` inputs to `forward` before it leads to errors down the line. I think a great place to do this would be in `t_batch_mode_transform`. I don't think any acquisition function supports `NaN` inputs, so it should be ok to simply raise the error before the `forward` call is evaluated. If this sounds like a good idea, I could add this to #577 or make a separate PR for it.

### PR DESCRIPTION
**Pull request by saitcakmak**
## Motivation

Closes #685. Implements simple checks in `gen_candidates_scipy` to catch `NaN` inputs and gradients, making debugging easier.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/pytorch/botorch/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Added unit tests.

### CODE REVIEW

File: botorch/generation/gen.py [hunk 1]
```diff
@@ -109,6 +114,16 @@ def f(x):
...
         gradf = _arrayify(torch.autograd.grad(loss, X)[0].contiguous().view(-1))
+        if np.isnan(gradf).any():
+            raise RuntimeError(
+                f"{np.isnan(gradf).sum()} elements of the {x.shape} gradient tensor "
+                "`gradf` are NaN."
+                + (
+                    "Consider using `dtype=torch.double`."
+                    if initial_conditions.dtype != torch.double
+                    else ""
+                )
```
Comment (Balandat) [inline botorch/generation/gen.py:125]: Let's make this a bit more readable

```suggestion
            msg = (
                f"{np.isnan(gradf).sum()} elements of the {x.shape} gradient tensor "
                " are NaN. This often indicates numerical issues."
            )
            if initial_conditions.dtype != torch.double:
                msg += " Consider using `dtype=torch.double`."
            raise RuntimeError(msg)
```

File: test/generation/test_gen.py [hunk 2]
```diff
@@ -138,6 +139,31 @@ def test_gen_candidates_torch_with_fixed_features(self):
...
+            (torch.double, "gradient tensor"),
+        ]:
+            ckwargs = {"dtype": dtype, "device": self.device}
+
+            test_ics = torch.rand(3, 1, **ckwargs)
+            test_grad = torch.tensor([0.5, 0.2, float("nan")], **ckwargs)
+            # test NaN in grad
+            with mock.patch("torch.autograd.grad", return_value=[test_grad]):
+                with self.assertRaisesRegex(RuntimeError, expected_regex):
+                    _ = gen_candidates_scipy(
```
Comment (Balandat) [inline test/generation/test_gen.py:154]:
```suggestion
                    gen_candidates_scipy(
```

File: test/generation/test_gen.py [hunk 3]
```diff
@@ -138,6 +139,31 @@ def test_gen_candidates_torch_with_fixed_features(self):
...
+                        initial_conditions=test_ics,
+                        acquisition_function=mock.Mock(return_value=test_ics),
+                    )
+
+            # test NaN in `x`
+            test_ics[-1] = torch.tensor(float("nan"))
+            with self.assertRaisesRegex(RuntimeError, "tensor `x` are NaN."):
+                _ = gen_candidates_scipy(
```
Comment (Balandat) [inline test/generation/test_gen.py:162]:
```suggestion
                gen_candidates_scipy(
```

**Review by Balandat — APPROVED** [review]
This is great, thanks also for the solid tests.

File: botorch/generation/gen.py [hunk 4]
```diff
@@ -98,6 +99,10 @@ def gen_candidates_scipy(
     )
 
     def f(x):
+        if np.isnan(x).any():
+            raise RuntimeError(
+                f"{np.isnan(x).sum()} elements of the {x.shape} tensor `x` are NaN."
```
Comment (sdaulton) [inline botorch/generation/gen.py:104]:
```suggestion
                f"{np.isnan(x).sum()} elements of the {x.size} array `x` are NaN."
```

This array is 1-d and kind of hard to interpret, so I think it is cleaner to just say x/N elements are NaN

File: botorch/generation/gen.py [hunk 5]
```diff
@@ -109,6 +114,14 @@ def f(x):
         loss = -acquisition_function(X_fix).sum()
         # compute gradient w.r.t. the inputs (does not accumulate in leaves)
         gradf = _arrayify(torch.autograd.grad(loss, X)[0].contiguous().view(-1))
+        if np.isnan(gradf).any():
+            msg = (
+                f"{np.isnan(gradf).sum()} elements of the {x.shape} gradient tensor "
```
Comment (sdaulton) [inline botorch/generation/gen.py:119]:
```suggestion
                f"{np.isnan(gradf).sum()} elements of the {x.size} gradient array "
```

same comment as above

**Pull request closed without merging**